File: desktop-prismqml/melodex_desktop/collection_controller.py
# coding: utf-8
# SPDX-License-Identifier: AGPL-3.0-only
"""QML-facing collection state backed by Melodex's existing JSON routes."""

from __future__ import annotations

from typing import Any

from PySide6.QtCore import QObject, Property, Signal, Slot

from .api_client import ApiClient, encoded_query, normalize_song, song_query


def normalize_collection(value: dict[str, Any]) -> dict[str, Any]:
    """Normalize collection metadata without changing backend semantics."""

    collection = value or {}
    kind = str(collection.get("kind") or "manual").strip().lower()
    source = str(collection.get("source") or "local").strip().lower()
    return {
        **collection,
        "id": str(collection.get("id") or ""),
        "name": str(collection.get("name") or "未命名歌单").strip(),
        "description": str(collection.get("description") or "").strip(),
        "cover": str(collection.get("cover") or "").strip(),
        "kind": kind,
        "content_type": str(
            collection.get("content_type") or "playlist"
        ).strip().lower(),
        "source": source,
        "creator": str(collection.get("creator") or "").strip(),
        "track_count": int(collection.get("track_count") or 0),
    }


def song_write_payload(value: dict[str, Any]) -> dict[str, Any]:
    """Build the exact write contract accepted by /collections/:id/songs."""

    song = normalize_song(value)
    return {
        "id": song["id"],
        "source": song["source"],
        "name": song["name"],
        "artist": song["artist"],
        "album": song["album"],
        "album_id": str(value.get("album_id") or value.get("albumId") or ""),
        "cover": song["cover"],
        "duration": max(0, int(song["duration"])),
        "extra": song["extra"],
    }


class CollectionController(QObject):
    """Own collection lists, selection and add-to-playlist operations."""

    collectionsChanged = Signal()
    selectedCollectionChanged = Signal()
    selectedIndexChanged = Signal()
    songsChanged = Signal()
    writableCollectionsChanged = Signal()
    targetIndexChanged = Signal()
    busyChanged = Signal()
    errorChanged = Signal()
    noticeChanged = Signal()
    collectionCreated = Signal(str)

    def __init__(self, api: ApiClient, parent: QObject | None = None) -> None:
        super().__init__(parent)
        self._api = api
        self._collections: list[dict[str, Any]] = []
        self._selected_collection: dict[str, Any] = {}
        self._songs: list[dict[str, Any]] = []
        self._target_id = ""
        self._preferred_collection_id = ""
        self._active_requests: set[str] = set()
        self._request_token_serial = 0
        self._error = ""
        self._notice = ""
        self._session_generation = 0
        self._list_serial = 0
        self._songs_serial = 0
        self._api.authenticatedChanged.connect(self._on_authentication_changed)
        if self._api.authenticated:
            self.refresh()

    def _begin_request(self, scope: str, replace: bool = True) -> str:
        was_busy = bool(self._active_requests)
        if replace:
            prefix = scope + ":"
            self._active_requests = {
                token for token in self._active_requests if not token.startswith(prefix)
            }
        self._request_token_serial += 1
        token = f"{scope}:{self._request_token_serial}"
        self._active_requests.add(token)
        if not was_busy:
            self.busyChanged.emit()
        return token

    def _end_request(self, token: str) -> None:
        was_busy = bool(self._active_requests)
        self._active_requests.discard(token)
        if was_busy and not self._active_requests:
            self.busyChanged.emit()

    def _set_error(self, message: str) -> None:
        if message != self._error:
            self._error = message
            self.errorChanged.emit()

    def _set_notice(self, message: str) -> None:
        if message != self._notice:
            self._notice = message
            self.noticeChanged.emit()

    def _clear_state(self) -> None:
        self._collections = []
        self._selected_collection = {}
        self._songs = []
        self._target_id = ""
        self._preferred_collection_id = ""
        self._active_requests.clear()
        self._error = ""
        self._notice = ""
        self.collectionsChanged.emit()
        self.selectedCollectionChanged.emit()
        self.selectedIndexChanged.emit()
        self.songsChanged.emit()
        self.writableCollectionsChanged.emit()
        self.targetIndexChanged.emit()
        self.busyChanged.emit()
        self.errorChanged.emit()
        self.noticeChanged.emit()

    @Slot()
    def _on_authentication_changed(self) -> None:
        self._session_generation += 1
        self._list_serial += 1
        self._songs_serial += 1
        self._clear_state()
        if self._api.authenticated:
            self.refresh()

    @Slot()
    def refresh(self) -> None:
        if not self._api.authenticated:
            return
        self._list_serial += 1
        serial = self._list_serial
        generation = self._session_generation
        previous_id = self._preferred_collection_id or str(
            self._selected_collection.get("id") or ""
        )
        self._preferred_collection_id = ""
        self._set_error("")
        request_token = self._begin_request("collections")

        def completed(payload, error: str, _status: int) -> None:
            self._end_request(request_token)
            if generation != self._session_generation or serial != self._list_serial:
                return
            if error:
                self._set_error(error)
                return
            previous_target_index = self.get_target_index()
            values = payload if isinstance(payload, list) else []
            self._collections = [
                normalize_collection(item) for item in values if isinstance(item, dict)
            ]
            self.collectionsChanged.emit()
            self.writableCollectionsChanged.emit()
            self._sync_target_collection(previous_target_index)
            selected_index = self._collection_index(previous_id)
            if selected_index < 0 and self._collections:
                selected_index = next(
                    (
                        index
                        for index, item in enumerate(self._collections)
                        if item["kind"] == "favorite"
                    ),
                    0,
                )
            self._select_index(selected_index)
            self.selectedIndexChanged.emit()

        self._api.request_json(
            "GET", "/music/collections?include_imported=1", completed
        )

    def _collection_index(self, collection_id: str) -> int:
        return next(
            (
                index
                for index, item in enumerate(self._collections)
                if item["id"] == collection_id
            ),
            -1,
        )

    def _writable_collections(self) -> list[dict[str, Any]]:
        return [item for item in self._collections if item["kind"] != "imported"]

    def _sync_target_collection(self, previous_index: int) -> None:
        writable = self._writable_collections()
        if not any(item["id"] == self._target_id for item in writable):
            favorite = next(
                (item for item in writable if item["kind"] == "favorite"), None
            )
            self._target_id = str((favorite or (writable[0] if writable else {})).get("id") or "")
        if previous_index != self.get_target_index():
            self.targetIndexChanged.emit()

    @Slot(int)
    def selectCollectionIndex(self, index: int) -> None:
        self._select_index(index)

    def _select_index(self, index: int) -> None:
        next_collection = (
            dict(self._collections[index])
            if 0 <= index < len(self._collections)
            else {}
        )
        if next_collection != self._selected_collection:
            self._selected_collection = next_collection
            self.selectedCollectionChanged.emit()
            self.selectedIndexChanged.emit()
        self._load_selected_songs()

    @Slot()
    def refreshSongs(self) -> None:
        self._load_selected_songs()

    def _load_selected_songs(self) -> None:
        self._songs_serial += 1
        serial = self._songs_serial
        generation = self._session_generation
        collection_id = str(self._selected_collection.get("id") or "")
        self._songs = []
        self.songsChanged.emit()
        if not collection_id:
            return
        self._set_error("")
        request_token = self._begin_request("songs")

        def completed(payload, error: str, _status: int) -> None:
            self._end_request(request_token)
            if generation != self._session_generation or serial != self._songs_serial:
                return
            if collection_id != str(self._selected_collection.get("id") or ""):
                return
            if error:
                self._set_error(error)
                return
            values = payload.get("songs", []) if isinstance(payload, dict) else payload
            if not isinstance(values, list):
                values = []
            self._songs = [
                normalize_song(item) for item in values if isinstance(item, dict)
            ]
            self.songsChanged.emit()

        self._api.request_json(
            "GET", f"/music/collections/{collection_id}/songs", completed
        )

    @Slot(str)
    def createCollection(self, name: str) -> None:
        name = name.strip()
        if not name:
            self._set_error("请输入歌单名称")
            return
        generation = self._session_generation
        self._set_error("")
        self._set_notice("")
        request_token = self._begin_request("create", replace=False)

        def completed(payload, error: str, _status: int) -> None:
            self._end_request(request_token)
            if generation != self._session_generation:
                return
            if error:
                self._set_error(error)
                return
            collection_id = str((payload or {}).get("id") or "")
            self._preferred_collection_id = collection_id
            self._set_notice(f"已创建歌单「{name}」")
            self.collectionCreated.emit(collection_id)
            self.refresh()

        self._api.request_json(
            "POST", "/music/collections", completed, {"name": name}
        )

    @Slot(int)
    def setTargetCollectionIndex(self, index: int) -> None:
        writable = self._writable_collections()
        if not 0 <= index < len(writable):
            return
        next_id = writable[index]["id"]
        if next_id != self._target_id:
            self._target_id = next_id
            self.targetIndexChanged.emit()

    @Slot("QVariantMap")
    def addSong(self, song: dict[str, Any]) -> None:
        target = next(
            (item for item in self._writable_collections() if item["id"] == self._target_id),
            None,
        )
        if target is None:
            self._set_error("请先创建或选择一个可写歌单")
            return
        payload = song_write_payload(song)
        if not payload["id"] or not payload["source"]:
            self._set_error("歌曲缺少来源标识，无法加入歌单")
            return
        generation = self._session_generation
        target_id = target["id"]
        target_name = target["name"]
        self._set_error("")
        self._set_notice("")
        request_token = self._begin_request("add", replace=False)

        def completed(_payload, error: str, _status: int) -> None:
            self._end_request(request_token)
            if generation != self._session_generation:
                return
            if error:
                self._set_error(error)
                return
            self._set_notice(f"已加入「{target_name}」")
            self._download_added_song(song)
            if target_id == str(self._selected_collection.get("id") or ""):
                self._load_selected_songs()

        self._api.request_json(
            "POST", f"/music/collections/{target_id}/songs", completed, payload
        )

    def _download_added_song(self, song: dict[str, Any]) -> None:
        path = "/music/download?" + encoded_query(
            song_query(song, embed="1", save_local="1")
        )

        def completed(_payload, error: str, _status: int) -> None:
            if error:
                print(f"[WARN] 歌曲已加入歌单，但后台下载到服务器失败：{error}")

        self._api.request_json("POST", path, completed)

    @Slot()
    def clearMessages(self) -> None:
        self._set_error("")
        self._set_notice("")

    def get_collections(self) -> list[dict[str, Any]]:
        return self._collections

    def get_selected_collection(self) -> dict[str, Any]:
        return self._selected_collection

    def get_selected_index(self) -> int:
        return self._collection_index(str(self._selected_collection.get("id") or ""))

    def get_songs(self) -> list[dict[str, Any]]:
        return self._songs

    def get_writable_collection_names(self) -> list[str]:
        return [item["name"] for item in self._writable_collections()]

    def get_target_index(self) -> int:
        return next(
            (
                index
                for index, item in enumerate(self._writable_collections())
                if item["id"] == self._target_id
            ),
            -1,
        )

    def get_busy(self) -> bool:
        return bool(self._active_requests)

    def get_error(self) -> str:
        return self._error

    def get_notice(self) -> str:
        return self._notice

    collections = Property("QVariantList", get_collections, notify=collectionsChanged)
    selectedCollection = Property(
        "QVariantMap", get_selected_collection, notify=selectedCollectionChanged
    )
    selectedIndex = Property(int, get_selected_index, notify=selectedIndexChanged)
    songs = Property("QVariantList", get_songs, notify=songsChanged)
    writableCollectionNames = Property(
        "QVariantList",
        get_writable_collection_names,
        notify=writableCollectionsChanged,
    )
    targetIndex = Property(int, get_target_index, notify=targetIndexChanged)
    busy = Property(bool, get_busy, notify=busyChanged)
    error = Property(str, get_error, notify=errorChanged)
    notice = Property(str, get_notice, notify=noticeChanged)

File: desktop-prismqml/cpp/CollectionController.cpp
#include "melodex/CollectionController.h"

#include "melodex/ApiClient.h"
#include "melodex/JsonUtils.h"

#include <QUrl>
#include <algorithm>

namespace melodex {
namespace {

QString encodedPathId(const QString &value) {
    return QString::fromLatin1(QUrl::toPercentEncoding(value));
}

}  // namespace

CollectionController::CollectionController(ApiClient *api, QObject *parent)
    : QObject(parent), m_api(api) {
    connect(m_api, &ApiClient::authenticatedChanged, this,
            &CollectionController::onAuthenticationChanged);
    if (m_api->authenticated())
        refresh();
}

QVariantList CollectionController::collections() const {
    return toVariantList(m_collections);
}

QVariantList CollectionController::songs() const { return toVariantList(m_songs); }

int CollectionController::selectedIndex() const {
    return collectionIndex(m_selectedCollection.value(QStringLiteral("id")).toString());
}

QStringList CollectionController::writableCollectionNames() const {
    QStringList names;
    for (const QVariantMap &collection : writableCollections())
        names.append(collection.value(QStringLiteral("name")).toString());
    return names;
}

int CollectionController::targetIndex() const {
    const QList<QVariantMap> writable = writableCollections();
    for (int index = 0; index < writable.size(); ++index) {
        if (writable.at(index).value(QStringLiteral("id")).toString() == m_targetId)
            return index;
    }
    return -1;
}

QString CollectionController::beginRequest(const QString &scope, bool replace) {
    const bool wasBusy = busy();
    if (replace) {
        const QString prefix = scope + QLatin1Char(':');
        for (auto it = m_activeRequests.begin(); it != m_activeRequests.end();) {
            if (it->startsWith(prefix))
                it = m_activeRequests.erase(it);
            else
                ++it;
        }
    }
    const QString token = scope + QLatin1Char(':') +
                          QString::number(++m_requestTokenSerial);
    m_activeRequests.insert(token);
    if (!wasBusy)
        emit busyChanged();
    return token;
}

void CollectionController::endRequest(const QString &token) {
    const bool wasBusy = busy();
    m_activeRequests.remove(token);
    if (wasBusy && !busy())
        emit busyChanged();
}

void CollectionController::setError(const QString &message) {
    if (message == m_error)
        return;
    m_error = message;
    emit errorChanged();
}

void CollectionController::setNotice(const QString &message) {
    if (message == m_notice)
        return;
    m_notice = message;
    emit noticeChanged();
}

void CollectionController::clearState() {
    m_collections.clear();
    m_selectedCollection.clear();
    m_songs.clear();
    m_targetId.clear();
    m_preferredCollectionId.clear();
    m_activeRequests.clear();
    m_error.clear();
    m_notice.clear();
    emit collectionsChanged();
    emit selectedCollectionChanged();
    emit selectedIndexChanged();
    emit songsChanged();
    emit writableCollectionsChanged();
    emit targetIndexChanged();
    emit busyChanged();
    emit errorChanged();
    emit noticeChanged();
}

void CollectionController::onAuthenticationChanged() {
    ++m_sessionGeneration;
    ++m_listSerial;
    ++m_songsSerial;
    clearState();
    if (m_api->authenticated())
        refresh();
}

void CollectionController::refresh() {
    if (!m_api->authenticated())
        return;
    const quint64 serial = ++m_listSerial;
    const quint64 generation = m_sessionGeneration;
    const QString previousId = !m_preferredCollectionId.isEmpty()
                                   ? m_preferredCollectionId
                                   : m_selectedCollection.value(QStringLiteral("id"))
                                         .toString();
    m_preferredCollectionId.clear();
    setError({});
    const QString token = beginRequest(QStringLiteral("collections"));
    m_api->requestJson(
        "GET", QStringLiteral("/music/collections?include_imported=1"),
        [this, token, generation, serial, previousId](const QVariant &payload,
                                                      const QString &error, int) {
            endRequest(token);
            if (generation != m_sessionGeneration || serial != m_listSerial)
                return;
            if (!error.isEmpty()) {
                setError(error);
                return;
            }
            const int previousTargetIndex = targetIndex();
            m_collections.clear();
            for (const QVariantMap &value : variantMaps(payload))
                m_collections.append(normalizeCollection(value));
            emit collectionsChanged();
            emit writableCollectionsChanged();
            syncTargetCollection(previousTargetIndex);
            int index = collectionIndex(previousId);
            if (index < 0 && !m_collections.isEmpty()) {
                index = 0;
                for (int candidate = 0; candidate < m_collections.size(); ++candidate) {
                    if (m_collections.at(candidate).value(QStringLiteral("kind")).toString() ==
                        QStringLiteral("favorite")) {
                        index = candidate;
                        break;
                    }
                }
            }
            selectIndex(index);
            emit selectedIndexChanged();
        });
}

int CollectionController::collectionIndex(const QString &collectionId) const {
    for (int index = 0; index < m_collections.size(); ++index) {
        if (m_collections.at(index).value(QStringLiteral("id")).toString() == collectionId)
            return index;
    }
    return -1;
}

QList<QVariantMap> CollectionController::writableCollections() const {
    QList<QVariantMap> writable;
    std::copy_if(m_collections.cbegin(), m_collections.cend(),
                 std::back_inserter(writable), [](const QVariantMap &item) {
                     return item.value(QStringLiteral("kind")).toString() !=
                            QStringLiteral("imported");
                 });
    return writable;
}

void CollectionController::syncTargetCollection(int previousIndex) {
    const QList<QVariantMap> writable = writableCollections();
    const bool targetExists = std::any_of(
        writable.cbegin(), writable.cend(), [this](const QVariantMap &item) {
            return item.value(QStringLiteral("id")).toString() == m_targetId;
        });
    if (!targetExists) {
        QVariantMap nextTarget;
        for (const QVariantMap &item : writable) {
            if (item.value(QStringLiteral("kind")).toString() == QStringLiteral("favorite")) {
                nextTarget = item;
                break;
            }
        }
        if (nextTarget.isEmpty() && !writable.isEmpty())
            nextTarget = writable.front();
        m_targetId = nextTarget.value(QStringLiteral("id")).toString();
    }
    if (previousIndex != targetIndex())
        emit targetIndexChanged();
}

void CollectionController::selectCollectionIndex(int index) { selectIndex(index); }

void CollectionController::selectIndex(int index) {
    const QVariantMap next = index >= 0 && index < m_collections.size()
                                 ? m_collections.at(index)
                                 : QVariantMap{};
    if (next != m_selectedCollection) {
        m_selectedCollection = next;
        emit selectedCollectionChanged();
        emit selectedIndexChanged();
    }
    loadSelectedSongs();
}

void CollectionController::refreshSongs() { loadSelectedSongs(); }

void CollectionController::loadSelectedSongs() {
    const quint64 serial = ++m_songsSerial;
    const quint64 generation = m_sessionGeneration;
    const QString collectionId = m_selectedCollection.value(QStringLiteral("id")).toString();
    m_songs.clear();
    emit songsChanged();
    if (collectionId.isEmpty())
        return;
    setError({});
    const QString token = beginRequest(QStringLiteral("songs"));
    const QString path = QStringLiteral("/music/collections/") + encodedPathId(collectionId) +
                         QStringLiteral("/songs");
    m_api->requestJson("GET", path,
                       [this, token, generation, serial, collectionId](
                           const QVariant &payload, const QString &error, int) {
                           endRequest(token);
                           if (generation != m_sessionGeneration || serial != m_songsSerial ||
                               collectionId != m_selectedCollection
                                                   .value(QStringLiteral("id")).toString())
                               return;
                           if (!error.isEmpty()) {
                               setError(error);
                               return;
                           }
                           QVariant values = payload;
                           if (payload.canConvert<QVariantMap>())
                               values = payload.toMap().value(QStringLiteral("songs"));
                           m_songs.clear();
                           for (const QVariantMap &song : variantMaps(values))
                               m_songs.append(normalizeSong(song));
                           emit songsChanged();
                       });
}

void CollectionController::createCollection(const QString &rawName) {
    const QString name = rawName.trimmed();
    if (name.isEmpty()) {
        setError(QStringLiteral("请输入歌单名称"));
        return;
    }
    const quint64 generation = m_sessionGeneration;
    setError({});
    setNotice({});
    const QString token = beginRequest(QStringLiteral("create"), false);
    m_api->requestJson(
        "POST", QStringLiteral("/music/collections"),
        [this, token, generation, name](const QVariant &payload,
                                        const QString &error, int) {
            endRequest(token);
            if (generation != m_sessionGeneration)
                return;
            if (!error.isEmpty()) {
                setError(error);
                return;
            }
            m_preferredCollectionId = payload.toMap().value(QStringLiteral("id")).toString();
            setNotice(QStringLiteral("已创建歌单「%1」").arg(name));
            emit collectionCreated(m_preferredCollectionId);
            refresh();
        },
        {{QStringLiteral("name"), name}});
}

void CollectionController::setTargetCollectionIndex(int index) {
    const QList<QVariantMap> writable = writableCollections();
    if (index < 0 || index >= writable.size())
        return;
    const QString nextId = writable.at(index).value(QStringLiteral("id")).toString();
    if (nextId == m_targetId)
        return;
    m_targetId = nextId;
    emit targetIndexChanged();
}

void CollectionController::addSong(const QVariantMap &song) {
    QVariantMap target;
    for (const QVariantMap &item : writableCollections()) {
        if (item.value(QStringLiteral("id")).toString() == m_targetId) {
            target = item;
            break;
        }
    }
    if (target.isEmpty()) {
        setError(QStringLiteral("请先创建或选择一个可写歌单"));
        return;
    }
    const QVariantMap payload = songWritePayload(song);
    if (payload.value(QStringLiteral("id")).toString().isEmpty() ||
        payload.value(QStringLiteral("source")).toString().isEmpty()) {
        setError(QStringLiteral("歌曲缺少来源标识，无法加入歌单"));
        return;
    }
    const quint64 generation = m_sessionGeneration;
    const QString targetId = target.value(QStringLiteral("id")).toString();
    const QString targetName = target.value(QStringLiteral("name")).toString();
    setError({});
    setNotice({});
    const QString token = beginRequest(QStringLiteral("add"), false);
    const QString path = QStringLiteral("/music/collections/") + encodedPathId(targetId) +
                         QStringLiteral("/songs");
    m_api->requestJson(
        "POST", path,
        [this, token, generation, targetId, targetName, song](
            const QVariant &, const QString &error, int) {
            endRequest(token);
            if (generation != m_sessionGeneration)
                return;
            if (!error.isEmpty()) {
                setError(error);
                return;
            }
            setNotice(QStringLiteral("已加入「%1」").arg(targetName));
            downloadAddedSong(song);
            if (targetId == m_selectedCollection.value(QStringLiteral("id")).toString())
                loadSelectedSongs();
        },
        payload);
}

void CollectionController::downloadAddedSong(const QVariantMap &song) {
    const QString query = encodedQuery(songQuery(
        song, {{QStringLiteral("embed"), QStringLiteral("1")},
               {QStringLiteral("save_local"), QStringLiteral("1")}}));
    m_api->requestJson(
        "POST", QStringLiteral("/music/download?") + query,
        [](const QVariant &, const QString &error, int) {
            if (!error.isEmpty())
                qWarning().noquote()
                    << "[WARN] 歌曲已加入歌单，但后台下载到服务器失败：" << error;
        });
}

void CollectionController::clearMessages() {
    setError({});
    setNotice({});
}

}  // namespace melodex
